modules/nlu/src/backend/tools/five-fold.ts

```diff
@@ -58,11 +61,28 @@ export class FiveFolder<T> {
       confusions: new Proxy({}, MapProxy)
     }
 
-    const shuffled = _.shuffle(this.dataset)
-    const chunks = _.chunk(shuffled, Math.ceil(shuffled.length / this.N))
-
-    await Promise.mapSeries(chunks, async testSet => {
-      const trainSet = _.flatten(chunks.filter(c => c !== testSet))
+    const testSets = _.chain(this.dataset)
```

Comment:
I think that works, but I can't say for sure just by reading it because it's hard to follow the logic

IMO something more procedural/verbose like this would make it simpler to read and debug for people coming across this piece of code

Moreover this class should remain generic, i.e. no reference to intents or utterances, as we will likely use it for more than intent classification

```js
/* If we take the analogy of Playing Cards
   `folds` represent card players
   we want to distribute each player cards of the same number fairly
   i.e. we want players to have the same amount of A's, K's and Q's
   but we want the color of the card to be distributed randomly */
const folds = new Array<T[]>(this.N)

// `this.dataset` -> we have been given 13 packs (A,2,3,4,5,6 etc..) of 4 cards (H,S,D,C)
for (let group of this.dataset) {
  const samples = _.shuffle(group) // we shuffle H,S,D,C
  const startPlayer = _.random(0, this.N, false) // we start dealing cards to a random player, so that the first player doesn't get more cards than the others
  for (let i = 0; i < group.length; i++) {
    folds[(i + startPlayer) % this.N].push(samples[i]) // we deal each player one of those card, so they get a random number of H,S,D,C
  }
}
```

Reply:
I didn't test the code above tbh, wrote it in github directly but hope you get the idea

Reply:
You are right, I was trying to avoid being statefull but this cleary makes this 10x more complicated and less readable.